src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,164 @@
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            addFirst(value);
+            return;
+        }
+        Node<T> newNode = new Node<>(last, value, null);
+        last.next = newNode;
+        last = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            addFirst(value);
+            return;
+        }
+        checkIndex(index - 1);
+        Node<T> current = getNodeByIndex(index);
+        Node<T> prevNode = current.prev;
+        Node<T> newNode = new Node<>(prevNode, value, current);
+        prevNode.next = newNode;
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeByIndex(index);
+        T deletedValue = current.item;
+        current.item = value;
+        return deletedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeByIndex(index);
+        if (size == 1) {
+            removeLast();
+            return current.item;
+        }
+        return unlink(current);
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> current = first;
+        while (current != null
+                && !(current.item == object
+                   || current.item != null && current.item.equals(object))) {
+            current = current.next;
+        }
+        return current != null
+                && (object == unlink(current) || object != null && object.equals(current.item));
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addFirst(T element) {
+        Node<T> current = first;
+        if (isEmpty()) {
+            Node<T> newNode = new Node<>(null, element, null);
+            first = newNode;
+            last = newNode;
+        } else {
+            first = new Node<>(null, element, current);
+        }
+        size++;
+    }
+
+    private void removeLast() {
+        first = null;
+        last = null;
+        size = 0;
+    }
+
+    private T unlink(Node<T> current) {
+        Node<T> nextNode = current.next;
+        Node<T> prevNode = current.prev;
+        size--;
+        if (size == 0) {
+            removeLast();
```

Comment:
do we need this first if?

Reply:
If Node(for delete) is the last, then prev and next are null. In other IF cases this situation is not checked.